chef_master/source/server_security.rst

```diff
@@ -310,6 +310,108 @@ These newer add-ons will also write all of their secrets to ``/etc/opscode/priva
 
+SSL Encryption Between Chef Infra Server and External PostgreSQL
+================================================================
+
+**New in Chef Server X.Y:**  Chef Server X.Y introduces the capability to encrypt traffic between Chef Infra Server and an external PostgreSQL configuration over SSL.  These instructions are not all-encompassing and assume familiarity with PostgreSQL administration, configuration, and troubleshooting.  Consult the appropriate PostgreSQL documentation as necessary, e.g. https://www.postgresql.org/docs/9.6/ssl-tcp.html.
+
+To enable this encryption (typical scenario):
+
+1. Ensure that OpenSSL is installed on the PostgreSQL machine (consult appropriate documentation as necessary, e.g. https://www.openssl.org/).
+
+2. Ensure that PostgreSQL has SSL support compiled-in (this applies whether you are compiling your own or using a pre-compiled binary).
+
+3. Place SSL certificates in the proper directories on the PostgreSQL machine, and ensure correct filenames, ownerships and permissions.
+
+4. Enable SSL support on PostgreSQL, and specify paths to the SSL certificates, if necessary.
+
+   Edit the postgresql.conf file on the PostgreSQL machine and ensure the relevant entries are present and not commented out (substitute the appropriate paths):
+
+.. code-block:: bash
+
+   # mandatory
+   ssl=on
+
+   # optional - uncomment accordingly
+   #ssl_cert_file='<PATH/TO/CERT/FILE>'
+   #ssl_key_file='<PATH/TO/KEY/FILE>'
+
+5. Consider forcing the use of SSL connections from the PostgreSQL side, otherwise non-SSL connections could be used.  This can be achieved by editing `pg_hba.conf` on the PostgreSQL machine, and changing the relevant Chef Server connections to `hostssl`.
+
+   Here is a sample `pg_hba.conf` file with `hostssl` connections for Chef Server (the contents of your `pg_hba.conf` will be different):
+
+.. code-block:: bash
+
+   # "local" is for Unix domain socket connections only
+   local      all             all                                     peer
+
+   # IPv4 local connections:
+   hostssl    all             all             127.0.0.1/32            md5
+
+   # IPv6 local connections:
+   hostssl    all             all             ::1/128                 md5
+
+   # nonlocal connections
+   hostssl    all             all            192.168.33.100/32        md5
+
+6. Restart PostgreSQL.  This can typically be done with the following command on the PostgreSQL machine (substitute the appropriate path):
+
+.. code-block:: bash
+
+   $ /<PATH/TO/POSTGRESQL>/postgresql restart
+
+7. Edit /etc/opscode/chef-server.rb on the Chef Server machine and add the following line:
+
+.. code-block:: ruby
+
+   postgresql['sslmode']='require'
+
```

Comment:
We should document the procedure for adding custom certs to chef server, or at least incorporate it by reference. It's not uncommon for the postgres server to use self signed or otherwise custom SSL certs that aren't on the standard chain of trust, and chef server needs to know to accept those.

Reply:
https://github.com/chef/chef-server/issues/1850

Reply:
Question: Does chef server currently even know or care about postgresql certs in an external postgresql configuration?

Reply:
re: <<Make sure we distinguish the tasks (enable SSL, set up certs, etc) from the means (edit this file). We won't be able to cover (or even keep up with) the various ways of getting postgres, but if we are clear about the steps and their intent, hopefully users will be able to self help.>>

I tried to convey this by stating the task to be done first ("Restart PostgreSQL") followed by an example means ("This can typically be done with the following command on the PostgreSQL machine...").  Let me know if this can be done in a better way.